pkg/operator/staticpod/controller/prune/prune_controller.go

```diff
@@ -0,0 +1,296 @@
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/client-go/util/workqueue"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	"github.com/openshift/library-go/pkg/operator/events"
+	"github.com/openshift/library-go/pkg/operator/resource/resourceread"
+	"github.com/openshift/library-go/pkg/operator/staticpod/controller/common"
+)
+
+type PruneController struct {
+	targetNamespace, podResourcePrefix          string
+	failedRevisionLimit, succeededRevisionLimit int
+
+	// command is the string to use for the pruning pod command
+	command []string
+	// queue only ever has one item, but it has nice error handling backoff/retry semantics
+	queue workqueue.RateLimitingInterface
+	// prunerPodImageFn returns the image name for the installer pod
+	prunerPodImageFn func() string
+
+	operatorConfigClient common.OperatorClient
+
+	kubeClient    kubernetes.Interface
+	eventRecorder events.Recorder
+
+	dynamicClient      dynamic.Interface
+	operatorConfigGVR  schema.GroupVersionResource
+	operatorConfigName string
+}
+
+const (
+	pruneControllerWorkQueueKey = "key"
+	statusConfigMapName         = "revision-status"
+	defaultResourceDir          = "/etc/kubernetes/static-pod-resources"
+	prunerPodName               = "revision-pruner"
+)
+
+func NewPruneController(
+	targetNamespace string,
+	podResourcePrefix string,
+	command []string,
+	kubeClient kubernetes.Interface,
+	operatorConfigClient common.OperatorClient,
+	eventRecorder events.Recorder,
+	dynamicClient dynamic.Interface,
+	operatorConfigGVR schema.GroupVersionResource,
+	operatorConfigName string,
+) *PruneController {
+	c := &PruneController{
+		targetNamespace:        targetNamespace,
+		podResourcePrefix:      podResourcePrefix,
+		command:                command,
+		failedRevisionLimit:    0,
+		succeededRevisionLimit: 0,
+
+		operatorConfigClient: operatorConfigClient,
+		kubeClient:           kubeClient,
+		eventRecorder:        eventRecorder,
+
+		dynamicClient:      dynamicClient,
+		operatorConfigGVR:  operatorConfigGVR,
+		operatorConfigName: operatorConfigName,
+
+		queue:            workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "PruneController"),
+		prunerPodImageFn: getPrunerPodImageFromEnv,
+	}
+
+	operatorConfigClient.Informer().AddEventHandler(c.eventHandler())
+
+	return c
+}
+
+func (c *PruneController) syncRevisionLimits() error {
+	config, err := c.dynamicClient.Resource(c.operatorConfigGVR).Get(c.operatorConfigName, metav1.GetOptions{})
+	if err != nil {
+		return err
+	}
+
+	revisionLimits, ok, err := unstructured.NestedMap(config.Object, "spec", "operatorRevisionLimit")
+	if err != nil {
+		return err
+	}
+	if ok {
```

Comment:
nit: i would flip this... `if !ok { return nil }`. reason is to reduce the levels of nesting in code